Lotto1단계 리뷰 요청합니다!

nhs0912 opened the pull request:
로또 1단계 리뷰 요청합니다! 

역시나 하면서 쉽지가 않았네요 ㅠㅠ

nhs0912 committed:
feat: README.md 기능 구현 목록 작성 및 로또 판매원 돈과 로또 바꾸어주기 기능 구현 및 테스트 작성

nhs0912 committed:
feat: 로또 자동 생성 및 구매한 로또 번호 출력 기능 구현

nhs0912 committed:
feat: 로또 당첨번호 구현 및 테스트 완료

nhs0912 committed:
feat : 당첨 통계 기능 구현 및 테스트 완료

hhlyn0328 commented on src/main/java/lotto/domain/Ball.java:
```diff
@@ -0,0 +1,45 @@
+
+    private void checkValidNumber(int number) {
+        if (!isRangeBetweenOneAndFortyFive(number)) {
+            throw new MyException(MyErrorCode.INVALID_RANGE_LOTTO_NUMBER);
```
MyException 이라는 이름보다, 사용 목적이 더 나타나면 좋을 것 같아요!

hhlyn0328 commented on src/main/java/lotto/domain/Ball.java:
```diff
@@ -0,0 +1,45 @@
+    private static final int MAX_LOTTO_NUMBER = 45;
+    private int number;
+
+    public Ball(int number) {
```
원시값 포장 :+1:

hhlyn0328 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,65 @@
+import java.util.*;
+import java.util.stream.Collectors;
+
+public class Lotto {
```
일급컬렉션 사용 :+1:

hhlyn0328 commented on src/main/java/lotto/domain/LottoCreator.java:
```diff
@@ -0,0 +1,34 @@
+        return selectedLottoBalls;
+    }
+
+    private void initLottoBall() {
```
static 블락을 사용해서 초기화 할 수 있을 것 같아요!
이렇게 초기화 하는 경우, 다른 메서드도 정적 메서드로 변경할 수 있을 것 같은데, 
이 클래스가 객체생성이 되어야하는지 한 번 고민해봐도 좋을 것 같아요!

nhs0912 replied:
static block을 통해서 초기화 하면 되는군요. 
처음에 그렇게 설정한 이유는 테스트 하기 위해서 Override를 통해서 makeLotto번호를 강제로 지정할 수 있기 때문이었어요! 
하지만 테스트 코드를 봤는데 숫자는 중요하지 않고 개수사 6개인 것만 체크해도 괜찮을거라고 생각이 들었어요. 
그래서 혜린님 의견을 반영해보았습니다! 덕분에 이건 저에게 생소한 부분이었지만 시야가 넓어졌네요!

hhlyn0328 commented on src/main/java/lotto/LottoMain.java:
```diff
@@ -0,0 +1,32 @@
+        int countOfLotto = Seller.returnLotto(purchaseMoney);
+        OutputView.printLottoCount(countOfLotto);
+
+        List<Lotto> lottoList = lottoService.createLotto(countOfLotto, lottoCreator);
```
createLotto 가 Lottos 를 반환하지 않는 이유가 있을까요?

nhs0912 replied:
앗! 바로 Lottos로 변환하면 되는데 두번 일을 하게 만들었네요! 바로 반영하였습니다! :)

hhlyn0328 commented on src/main/java/lotto/view/OutputView.java:
```diff
@@ -0,0 +1,58 @@
+        System.out.println(countOfLotto + "개를 구매했습니다.");
+    }
+
+    public static void printLotto(List<List<Integer>> lottos) {
```
Lotto, Lottos의 toString을 재정의하고 활용해보면 어떨까요?

nhs0912 replied:
이렇게 재정의하면 훨씬 더 코드가 간편해졌네요! 덕분에 코드가 간결해지고 toString()을 사용해보는 공부를 하였습니다 :)

hhlyn0328 commented on src/main/java/lotto/domain/Ball.java:
```diff
@@ -0,0 +1,45 @@
+        return this.number;
+    }
+
+    @Override
```
저는 개인적으로 equals & hashCode 코드보다 비지니스로직을 담는 Private method를 더 상위에 두는 편입니다
아무래도 equals & hashCode 내부 로직보다 Private method 내부의 코드를 좀 더 중요하게(?) 봐야한다고 생각하기 때문인데요 
희석님은 이렇게 작성하신 이유가 있을까요?

nhs0912 replied:
object(ball)를 contains() method를 통해서 존재하는지 안하는지 판단하려고 override를 하였습니다. 
사실 혜린님 말씀대로 변경한거 같긴한데 제가 잘 이해했는지 모르겠네요! 

제가 이해하는게 내부 로직으로 하는것보다 private Method를 생성하여서 기능 구현을 하라는 말씀 맞을까요??

hhlyn0328 replied:
앗 아니요 클래스 내에 코드작성 순서를 말씀드린것이었어요

저는 변수선언, 생성자, public 메서드, private 메서드, 
단순 getter, toString, equals, hashCode 등.. 의 순서로 코드를 배치하는데요

제가 저러한 순서로 작성하는 이유는 다른 개발자가 코드를 파악할 때, 
유심히 봐야하는 비지니스 로직은 모여있는게 더 좋을 것 같다고 생각하기 때문입니다

아주 개인적인 선택인 부분이라, 희석님은 어떤 생각을 가지고 계신지 궁금해서 문의드렸던 부분이었습니다!

nhs0912 replied:
아 질문을 이해 못해가지고 ㅠㅠㅠ 
다시 답변할게요! ㅋㅋㅋ 저도 일단은 제일 먼저 개발자는 생성자와 
외부로 노출된 method를 볼거 같아서 변수 선언 밑에 위치해요! 
그 다음에 private method를 위치하구요. 최대한 private은 숨겨야할거 같아서 
public 밑으로 합니다! 
단순 getter는 위에 변수로도 유추할수 있어서 거의 마지막에 하구요! 

근데 지금 이렇게 정리하다보니 문득 드는 생각이 public method안에 
각각 1개씩 또는 여러개 private method를 사용한다면 

private method의 동등한 위치에서 순서정리가 또 필요할지는 생각을 못해봤네요! 
혹시 거기까지 우선순위를 생각해놓으셨나요?? 그리고 제가 다니고 있는 직장에서는
사실상 이런 순서를 팀으로 정하거나 그러지 않는데 
혜린님의 회사에서는 이런 규칙도 정하는건가요? 

[ 코드 배치 순서 ]
상수 선언
변수 선언
셍성자
public method
private method
단순 getter

위와 같이 생각하고서 밑에 있는거는 사실 순서 상관없이 했던거 같아요! 
혹시 밑에 메서드 같은 경우도 순서가 있다면 더 좋을까요?? 

toString
equals
hashcode

hhlyn0328 replied:
`혜린님의 회사에서는 이런 규칙도 정하는건가요?`
아닙니다! ㅋㅋㅋ 

저는 리뷰하면서 많이 느끼고 배울 수 있고,
다양한 의견을 듣는 기회라고 생각하기 때문에 궁금해지면 큰 의미없이 질문을 남기는데요

저는 보통 equals, hashcode를 제일 마지막에 썼었거든요, 자동생성되는 코드이기도 하고 유심히 안본다고 생각해서요! 
그런데 희석님 코드를 보다가 비지니스 로직이 담긴 private method 앞에 equals, hashcode 가 위치해 있어서 
오? 하고 물어본 부분이었습니다! 
혹시 자동생성된 코드에서 변경된 내용이 있었는지, 좀 더 중요하게 생각하시는지, 다른 이유가 있는지? 해서요 :) 

지금 답변 주신 내용을 보니 제 생각과 크게 다르지 않으신거 같아, 좋은 답변이 되었습니다 :) 
다음 단계로 고고! ㅋㅋ

hhlyn0328 commented on src/main/java/lotto/domain/LottoCreator.java:
```diff
@@ -0,0 +1,34 @@
+    private static final int MIN_LOTTO_NUMBER = 1;
+    private static final int MAX_LOTTO_NUMBER = 45;
+
+    private static final List<Integer> lottoBalls = new ArrayList<>();
```
이 lottoBalls 는 변경되지 않아야하므로 static final 로 선언하신 것 같아요 
하지만 컬렉션은 이처럼 선언해도 수정이 가능합니다, unmodifiableCollection 에 대해 찾아보시면 좋을 것 같아요 !   

그리고 상수 선언 컨벤션에 맞도록, 대문자와 언더스코어를 사용해주시면 좋을 것 같습니다!

nhs0912 replied:
실수로 상수컨벤션을 안지켰네요 ㅠㅠㅠ
그리고 사실 컬렉션은 final로 해도 재할당이 안되는거지 값은 변한다는걸 알고 있는데도 실수 했네요 ㅠㅠ 
요구사항 반영하였습니다! :)

hhlyn0328 commented on src/test/java/lotto/domain/LottoTest.java:
```diff
@@ -0,0 +1,111 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
+
+class LottoTest {
```
테스트케이스를 꼼꼼히 잘 작성해주셨네요 :)

nhs0912 replied:
하나 질문이 있는데요 

- 
```
public boolean isEven(){
  for(Integer number : numbers){

   if(number % 2 == 0){
       return true;
   } 

   return false;
  }
}
```

이런 코드가 있을때 depth를 줄일수가 있을까요??

hhlyn0328 replied:
```
public boolean isAllEven(){
  for(Integer number : numbers){
     return isEven(number);
  }
}
public boolean isEven(){
  if(number % 2 == 0){
    return true;
  } 
  return false;
}  
```
코드 예시가 무조건 반복문을 1번밖에 수행하지 않아서 적절한지 모르겠지만 .. 
일반적으로 for문안에 if문이 있다고 가정했을 때,
저는 위처럼 내부 로직을 별도의 메서드로 분리할 것 같아요!

nhs0912 replied:
아! 제가 질문을 정확하게 잘 안했네요 ㅠㅠ 
```java
public Member findMember(List<Member> members, int grade){
  list<Member> findMembers = new ArrayList<>(); 
  for(Member member : members){
      if(member.getGrade == grade){
        findMembers.add(member);
      }
  }
  return fineMembers;  
}
```

위와 같은 코드를 제가 depth를 1로 해야한다고 봤을 때 

```java
public Member findMember(List<Member> members, int grade){
  list<Member> findMembers = new ArrayList<>(); 
  for(Member member : members){
      findMembers = matchMember(findMembers,member,grade);
  }
  return fineMembers;  
}

private list<Member> matchMember(list<Member> findMembers, Member member, int grade){  
    if(member.getGrade == grade){
      findMembers.add(member);
    }
    return findMembers;    
}

```

이런식으로 해야한다고 생각이 들거든요..이런것은 효율적인 방법일까 라는 생각을 하게 되었어요. 
그렇다고 밑에 코드처럼하면 더 안좋을거 같다라는 생각을 하고요! 람다로 할 수 밖에 없을까 다른 방법이 혹시 있을까 생각하는데 혜린님 생각은 어떠한지 궁금했어요! 

```java
public Member findMember(List<Member> members, int grade){
  list<Member> findMembers = new ArrayList<>(); 
  for(Member member : members){
      Member findMember = matchMember(member,grade);
      findMembers.add(findMember);
  }
  return fineMembers;  
}

private Member matchMember(Member member, int grade){  
    if(member.getGrade == grade){
      return member;
    }
    return null;    
}

```

hhlyn0328 replied:
오 이건 꽤 까다롭네요 🤔 
저는 먼저 stream().filter().collect() 가 가장 먼저 떠올랐습니다.. ㅋㅋ
다른 방법은.. true, false를 key로 갖고, Member List를 value로 갖는 맵에 일괄적으로 추가할 수도 있을 것 같고..
쥐어짜내면 어떻게든 되겠지만 굳이..? 싶어요 

제가 앞서 답변 드렸을 때는, 내부 로직이 분리가능한 독립적인 기능으로 봤는데 
리스트에 추가하는 건, 그렇다고 보기 어려울 것 같아서요
Stream() 짱 👍  ㅋㅋㅋㅋ 

그보다 member.getGrade == grade 비교보다는 
member에게 메세지를 던져보면 좋겠다 하는게 더 먼저 보이네요 😂 😂

hhlyn0328 reviewed (changes requested):
3단계 로또 자동 미션도 잘 진행해주셨네요 :) 

몇 가지 고민해볼 부분에 대해 피드백 남겼습니다
확인 부탁드리고, 의견도 자유롭게 남겨주세요!

nhs0912 committed:
refactor: 요구사항 반영
- toString으로 OutputView method Parameter 수정
- equals & hashCode 제거 및 private Method 정의
- static block 사용
- Exception class name 변경

nhs0912 committed:
refactor : 쓸모없는 주석 제거

hhlyn0328 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -29,14 +28,16 @@ public Lotto(List<Ball> balls) {
-        return this.balls().contains(ball);
+    private boolean isContain(Ball ball) {
+        return this.balls.stream()
+                .anyMatch(compareNumber -> ball.number() == compareNumber.number());
```
제가 괜한 질문을 해서, 객체 내부의 값을 꺼내서 비교하게 했네요 😢 
equals & hashcode 복원하셔서 기존 로직으로 써주셔도 될 것 같아요!

nhs0912 replied:
`제가 괜한 질문을 해서, 객체 내부의 값을 꺼내서 비교하게 했네요 😢`

아니에요 덕분에 equals & hashcode를 사용하지 않고 코딩하는 방법을 생각해본거 같아요! ㅋㅋㅋ

hhlyn0328 reviewed (approved):
요구사항 반영을 잘 해주셨네요 👍 👍 

다음 단계 진행을 위해 머지하겠습니다!!
질문에 대한 답변 코멘트에 남겨두었으니 확인 부탁드려요 :)
